Fix serialization of FQNs

hubertp opened the pull request:
### Pull Request Description

Serialization of FQNs' metadata was broken - we attempted to serialiaze concrete modules and that's against the design.

Added a test illustrating the problem which would previously fail during serialization.

Closes #5037 

### Important Notes

It is still hard to discover problems like this because SerializationManager creates system threads; when the exception occurs, it is typically not the real cause. Creating regular threads via `createThread` seems to be problematic for `native-image`. There is a simple workaround for the former but will leave it out for another PR to simplify the review.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Fix serialization of FQNs

Serialization of FQNs' metadata was broken - we attempted to serialiaze
concrete modules and that's against the design.

Added a test illustrating the problem which would previously fail during
serialization.

It is still hard to discover problems like this because
SerializationManager creates system threads; when the exception occurs
there it is typically not the real cause. Creating regular threads via
`createThread` seems to be problematic for `native-image`.
There is a simple workaround for the former but will leave it out for
another PR to simplify the review.

hubertp committed:
remove spurious file

hubertp committed:
fmt

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/compiler/SerializerTest.java:
```diff
@@ -0,0 +1,66 @@
+    assertEquals(result.compiledModules().exists(m -> m == module), true);
+    var serializationManager = new SerializationManager(ensoContext.getCompiler());
+    serializationManager.serialize(module, true);
+    Thread.sleep(1000); // Ensure module is serialized
```
Can we modify `serialize` to return a future and let the test wait for it? Something like:
```diff
diff --git a/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala b/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala
index 973c93a91..36ad0862a 100644
--- a/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala
@@ -15,6 +15,8 @@ import java.util.concurrent.{
 }
 import java.util.logging.Level
 import scala.collection.mutable
+import java.util.concurrent.Future
+import java.util.concurrent.CompletableFuture
 
 class SerializationManager(compiler: Compiler) {
 
@@ -79,10 +81,10 @@ class SerializationManager(compiler: Compiler) {
     * mutated beneath it.
     *
     * @param module the module to serialize
-    * @return `true` if `module` has been scheduled for serialization, `false`
+    * @return Future if `module` has been scheduled for serialization, `null`
     *         otherwise
     */
-  def serialize(module: Module, useGlobalCacheLocations: Boolean): Boolean = {
+  def serialize(module: Module, useGlobalCacheLocations: Boolean): Future[_] = {
     logger.log(
       debugLogLevel,
       s"Requesting serialization for module [${module.getName}]."
@@ -103,12 +105,11 @@ class SerializationManager(compiler: Compiler) {
     )
     if (compiler.context.getEnvironment.isCreateThreadAllowed) {
       isWaitingForSerialization.put(module.getName, task)
-      pool.execute(task)
+      pool.submit(task)
     } else {
       task.run()
+      CompletableFuture.completedFuture(true)
     }
-
-    true
   }
 
   /** Deserializes the requested module from the cache if possible.
```

hubertp replied:
I was hesitant to change the API of `SerializationManager` but sounds like a good idea.

hubertp committed:
Serialization now returns a Future

Now in tests we can wait for the Future to complete.
Had to add a bunch of synchronization points. I suspect in the past we
could have had potential race-conditions without them.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala:
```diff
@@ -82,7 +85,10 @@ class SerializationManager(compiler: Compiler) {
     * @return `true` if `module` has been scheduled for serialization, `false`
```
The `@return` tag description no longer fully match the return type.

hubertp committed:
update documentation

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/fqn-serialization-5037

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/fqn-serialization-5037
